### PR TITLE
Adding HAL profiling API and RenderDoc support for Vulkan.

### LINKED ISSUE #45: Add profiling support to the HAL API for use in benchmarks
We'll want a way to expose various metrics from the HAL implementations in a way that avoids excessive sequencer work. This could be accomplished by a begin/end profiling API and a resulting profile that contains cumulative, sampled, or averaged counter values per-backend. On Vulkan this may mean some vendor-specific performance counters in addition to timestamps inserted into command buffers.

### PR DESCRIPTION
**Pull request by benvanik**
TLDR: configure with `-DIREE_ENABLE_RENDERDOC_PROFILING=ON`, pass the `--device_profiling_mode=queue` flag to the IREE tools, and launch the tools from the RenderDoc UI in order to get a capture (or use `renderdoccmd capture`):
image

Things are set up to allow for other profiling modes in the future but how best to integrate those is TBD. We can figure out how to scale this with other tooling and on other backends but the rough shape of the API should be compatible with the various backend APIs we target (D3D/Metal/CUDA/Vulkan/perf/etc). Note that because RenderDoc will also capture D3D the cmake flag is generic but both the Vulkan and D3D HAL implementations will need to load it themselves (no real code worth sharing as D3D naturally only needs the Windows API query path).

Docs have notes that I've verified on Windows. Someone looking to use this on Android will need to figure that out and can add what they find.

Fixes #45. Forty five. Wow.

### CODE REVIEW

File: runtime/src/iree/hal/drivers/vulkan/CMakeLists.txt [hunk 1]
```diff
@@ -118,3 +118,12 @@ iree_cc_test(
...
 ### BAZEL_TO_CMAKE_PRESERVES_ALL_CONTENT_BELOW_THIS_LINE ###
+
+# If renderdoc support is enabled we can make use of it in the device.
+# Note that we disable this by default as it introduces a backdoor.
+if(IREE_ENABLE_RENDERDOC_PROFILING)
+  target_compile_definitions(iree_hal_drivers_vulkan_vulkan
+    PUBLIC
+      "IREE_HAL_VULKAN_HAVE_RENDERDOC=1"
+  )
+endif()
```
Comment (ScottTodd) [inline runtime/src/iree/hal/drivers/vulkan/CMakeLists.txt:129]: Thanks for making this CMake-only and optional. That will simplify pulling this into the internal Bazel repo.

(other context: https://github.com/iree-org/iree/pull/6207 removed our prior `third_party/renderdoc/` configuration)

File: runtime/src/iree/hal/drivers/vulkan/vulkan_device.cc [hunk 2]
```diff
@@ -36,6 +36,96 @@
...
+// Returns a handle to the RenderDoc API when it is hooking the process.
+// Returns NULL when RenderDoc is not present (or valid).
+static RENDERDOC_API_LATEST* iree_hal_vulkan_query_renderdoc_api(
+    VkInstance instance) {
+  pRENDERDOC_GetAPI RENDERDOC_GetAPI = NULL;
+#if defined(IREE_PLATFORM_WINDOWS)
+
+  // NOTE: RenderDoc only supports hooking so we can't use LoadLibrary - if
+  // we're going to use RenderDoc its library must already be loaded.
+  if (HMODULE hook_module = GetModuleHandleA("renderdoc.dll")) {
```
Comment (ScottTodd) [inline runtime/src/iree/hal/drivers/vulkan/vulkan_device.cc:64]: Is this DLL easy to discover this way? See https://github.com/iree-org/iree/pull/2131

Comment (benvanik) [reply runtime/src/iree/hal/drivers/vulkan/vulkan_device.cc:64]: It's not needed here as we aren't loading the library and instead just fetching the handle if already loaded. GetModuleHandleA scans a global flattened module list where paths don't mater. Loading the library after we've touched any Vulkan files won't work and we don't want to be doing this unconditionally (why it's unsupported by RenderDoc - same reason the posix path uses RTLD_NOLOAD). `renderdoccmd capture iree-run-module ...` can be used to capture from command lines (I'll add that to docs).

Comment (benvanik) [reply runtime/src/iree/hal/drivers/vulkan/vulkan_device.cc:64]: Added docs on `renderdoccmd capture`.

**Review by ScottTodd — APPROVED** [review]

**Review by antiagainst — APPROVED** [review]
Oh wow, thanks a lot for adding this! :D

**Comment by ScottTodd** [comment]
An unrelated test failed (flaked? timed out?)

https://github.com/iree-org/iree/actions/runs/3317175665/jobs/5480042484

```
ConnectionResetError: [Errno 104] Connection reset by peer

Failed Tests (1):
  TENSORFLOW_TESTS :: iree_tf_tests/uncategorized/llvmcpu__mobile_bert_squad.run\
```

I saw the same thing on https://github.com/iree-org/iree/pull/10871

**Comment by benvanik** [comment]
Ew yeah that's a tf flake downloading a file? Nasty.
